pymc/distributions/distribution.py

```diff
@@ -622,13 +685,15 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
+        def dist_moment(rv, size, *dist_params):
+            fgraph = get_rv_fgraph(dist, dist_params, size)
+            replace_moments = MomentRewrite()
```

Comment:
I am still worried that this will rewrite `dist_params` as well, and not just the CustomDist graph between rv and dist_params. 

Would the following work? Clone the inner fgraph from rv.owner.op.fgraph. This graph doesn't have `dist_params` directly but dummy placeholders called `NominalVariable`s. Apply the rewrite on this inner graph, and once you are done, replace the `NominalVariable`s by the respective `dist_params`?

Reply:
Actually what I am suggesting is pretty simple to what you did with `Scan`, so you could perhaps reuse some of the same logic?

Reply:
@ricardoV94  When I trying to create `FunctionGraph` with `NominalVariable`s I get error about missing input values - I think `FunctionGraph` constructor do not support dummy placeholders as input. But I`m try another approach:
```python
def dist_moment(rv, size, *dist_params, dist):
    rv = dist(*dist_params, size=size)
    inputs, outputs = dist_params, [rv.owner.out]
    fgraph_topo = io_toposort(inputs, outputs)
    replace_with_moment = []
    to_replace_set = set()

    for nd in fgraph_topo:
          if nd not in to_replace_set and isinstance(
               nd.op, (RandomVariable, SymbolicRandomVariable)
          ):
                replace_with_moment.append(nd.out)
                to_replace_set.add(nd)
     givens = {}
     for item in replace_with_moment:
            givens[item] = moment(item)
     [out] = clone_replace(outputs, replace=givens)
     return out
```
Looks like it work but do not support `Scan` for now (but I can add it). Do you think such approach will be better?

Reply:
You don't have to recreate the `FunctionGraph`, you can just do `fgraph.clone`?

Reply:
@ricardoV94 I try this approach, but I need to create `CustomDist` op first because just output of `dist` function do not have `fgraph` attribute.
```python
def dist_moment(rv, size, *dist_params):
            size = normalize_size_param(size)
            dummy_size_param = size.type()
            dummy_dist_params = [dist_param.type() for dist_param in dist_params]
            dummy_rv = dist(*dummy_dist_params, dummy_size_param)
            dummy_params = [dummy_size_param] + dummy_dist_params
            rv_type = type(
                class_name,
                (CustomSymbolicDistRV,),
                # If logp is not provided, we try to infer it from the dist graph
                dict(
                    inline_logprob=logp is None,
                ),
            )
            rv_op = rv_type(
                inputs=dummy_params,
                outputs=[dummy_rv],
                ndim_supp=ndim_supp,
            )
            fgraph = rv_op.fgraph.clone()
            replace_moments = MomentRewrite()
            replace_moments.rewrite(fgraph)
            for i, par in enumerate([size] + list(dist_params)):
                fgraph.replace(fgraph.inputs[i], par)
            [moment] = fgraph.outputs
            return moment
```
Do you think this way better? It is  also do not work with `Scan` yet. I need to figure out why and I will move graph creation logic to helper function if we will decide to keep this way.

Reply:
You have fgraph in the `rv.owner.op` from the `rv` that's passed as input.

About Scan what exactly is the problem? The approach you had written already seemed fine.  Was it failing?

Reply:
For scan I got Assertion Error:
```
        # allocate storage for intermediate computation
        for node in order:
            for r in node.inputs:
                if r not in storage_map:
>                   assert isinstance(r, Constant)
E                   AssertionError

../../miniconda3/envs/pymc-dev/lib/python3.11/site-packages/pytensor/link/utils.py:135: AssertionError
```
It`s hapens when I run `TestCustomSymbolicDist::test_custom_dist_default_moment_inner_graph` test case

Reply:
If I create rv like this it`s do not have fgraph atribute:
```
rv = dist(*dist_params, size=size)
>>> rv
Exp.0
>>> rv.owner
Exp(normal_rv{0, (0, 0), floatX, False}.out)
>>> rv.owner.op
Elemwise(scalar_op=exp,inplace_pattern=<frozendict {}>)
>>> rv.owner.op.fgraph
*** AttributeError: 'Elemwise' object has no attribute 'fgraph'
```
I use `lambda mu, sigma, size: pt.exp(pm.Normal.dist(mu, sigma, size=size))` for `dist` function  here.

Reply:
No, use the `rv` that is passed to the moment function directly, don't recreate it

Reply:
Ohm yeah, my bad. Will try it

Reply:
@ricardoV94 No, I can not use `rv` that is passes as parameter:
```
>>> rv
*** Not yet returned!
```
So I tried to re-create it